File: packages/python/modifyFilesMetadataInDirectory.py
import os
from pptx import Presentation


def modify_pptx_metadata(file_path):
    try:
        print(file_path)
        # 获取文件名（不包含扩展名）作为标题
        file_name = os.path.splitext(os.path.basename(file_path))[0]
        prs = Presentation(file_path)
        prs.core_properties.title = file_name
        # prs.core_properties.title = "晓橙"
        prs.core_properties.author = "晓橙"
        prs.core_properties.comments = "晓橙"
        prs.save(file_path)
    except Exception as e:
        print(f"Error occurred while updating metadata for {file_path}: {e}")


def modify_metadata_in_directory(directory_path):
    for root, dirs, files in os.walk(directory_path):
        for file_name in files:
            if file_name.endswith(".pptx"):
                file_path = os.path.join(root, file_name)
                modify_pptx_metadata(file_path)


# 示例用法
directory_path = "D:\\qrj\\download\\3.1"
modify_metadata_in_directory(directory_path)

File: packages/node/modifyFilesMetadataInDirectory.js
const fs = require('fs').promises;
const path = require('path');
const { exec } = require('child_process');
const iconv = require('iconv-lite');

async function modifyFilesMetadataInDirectory(
  directoryPath,
  title,
  author,
  comments
) {
  try {
    // 读取目录中所有文件和子目录
    const entries = await fs.readdir(directoryPath, { withFileTypes: true });

    // 遍历每个文件和子目录
    for (const entry of entries) {
      const entryPath = path.join(directoryPath, entry.name);
      if (entry.isDirectory()) {
        // 如果是子目录，则递归调用该函数
        await modifyFilesMetadataInDirectory(
          entryPath,
          title,
          author,
          comments
        );
      } else {
        // 如果是文件，则修改元数据
        await modifyFileMetadata(entryPath, title, author, comments);
      }
    }
  } catch (error) {
    console.error('Error occurred while modifying files metadata:', error);
  }
}

// async function modifyFileMetadata(filePath, title, author, comments) {
//   try {
//     // 获取文件的当前元数据
//     const stats = await fs.stat(filePath);

//     // 更新文件的标题、作者和备注
//     const updatedStats = {
//       ...stats,
//       title: title || '',
//       author: author || '',
//       comments: comments || ''
//     };

//     // 写入更新的元数据
//     await fs.utimes(filePath, updatedStats.atime, updatedStats.mtime);
//     console.log(
//       `File metadata updated successfully for ${path.basename(filePath)}`
//     );
//   } catch (error) {
//     console.error('Error occurred while updating file metadata:', error);
//   }
// }


function modifyFileMetadata(filePath, title, author, comments) {
  // filePath = encodePath(filePath)
  const command = `exiftool -Title="${title}" -Author="${author}" -Comment="${comments}" ${filePath}`;
  exec(command, { encoding: 'buffer' }, (error, stdout, stderr) => {
    if (error) {
      console.error(
        `Error occurred while modifying file metadata: ${error.message}`
      );
      return;
    }
    if (stderr) {
      console.error(`ExifTool stderr: ${stderr}`);
      return;
    }
    console.log(`File metadata updated successfully for ${filePath}`);
  });
}

// 编码文件路径以避免 FileName encoding not specified 错误
function encodePath(filePath) {
  return iconv.encode(filePath, 'win1252').toString();
}

// // 示例用法
// const filePath = 'path/to/your/file.txt';
// const title = 'New Title';
// const author = 'John Doe';
// const comments = 'This is a modified file.';

// modifyFileMetadata(filePath, title, author, comments);

module.exports = {
  modifyFilesMetadataInDirectory,
  modifyFileMetadata,
  encodePath
};
